adapter/internal/discovery/xds/server.go

```diff
@@ -405,23 +412,61 @@ func GenerateEnvoyResoucesForGateway(gatewayName string) ([]types.Resource,
 	}
 
 	envoyGatewayConfig, gwFound := gatewayLabelConfigMap[gatewayName]
-	listener := envoyGatewayConfig.listener
-	if !gwFound || listener == nil {
+	listeners := envoyGatewayConfig.listeners
+	if !gwFound || listeners == nil || len(listeners) == 0 {
 		return nil, nil, nil, nil, nil
 	}
-	routesFromListener := listenerToRouteArrayMap[listener.Name]
-	var vhostToRouteArrayFilteredMap = make(map[string][]*routev3.Route)
-	for vhost, routes := range vhostToRouteArrayMap {
-		if vhost == systemHost || checkRoutes(routes, routesFromListener) {
-			vhostToRouteArrayFilteredMap[vhost] = routes
+	routeConfigs := make([]*routev3.RouteConfiguration, 0)
+	for _, listener := range listeners {
+		for vhost, routes := range vhostToRouteArrayMap {
+			matchedListener, found := common.FindElement(data_holder.GetAllGatewayListeners(), func(listenerLocal gwapiv1b1.Listener) bool {
+				if (listenerLocal.Hostname != nil && common.MatchesHostname(vhost, string(*listenerLocal.Hostname))) {
+					if (listener.Name == common.GetEnvoyListenerName(string(listenerLocal.Protocol), uint32(listenerLocal.Port))) {
+						return true
+					}
+				}
+				return false
+			})
+			if found {
+				// Prepare the route config name based on the gateway listener section name.
+				routeConfigName := common.GetEnvoyRouteConfigName(listener.Name, string(matchedListener.Name));
+				routesConfig := oasParser.GetRouteConfigs(map[string][]*routev3.Route{vhost:routes}, routeConfigName, envoyGatewayConfig.customRateLimitPolicies)
+				
+				routeConfigMatched, alreadyExistsInRouteConfigList := common.FindElement(routeConfigs, func(routeConf *routev3.RouteConfiguration) bool {
+					if (routeConf.Name == routesConfig.Name) {
+						return true
+					}
+					return false
+				})
+				if alreadyExistsInRouteConfigList {
+					logger.LoggerAPKOperator.Debugf("Route already exists. %+v", routesConfig.Name)
```

Comment:
Can this happen? I didn't get how there could be duplicate routes at this level.